service/dbfs/utilities.go

```diff
@@ -11,116 +11,261 @@ import (
-// This can be used with [io.Copy] to maximize throughput, as
-// it uses the maximum buffer size allowed by the DBFS API.
-func (r *FileReader) WriteTo(w io.Writer) (n int64, err error) {
-	buf := make([]byte, maxDbfsReadSize)
-	nwritten := int64(0)
-	for {
-		n, err := r.Read(buf)
-		if err != nil {
-			// EOF on read means we're done.
-			// For writers being done means returning a nil error.
-			if err == io.EOF {
-				err = nil
-			}
-			return nwritten, err
-		}
-		n64, err := io.Copy(w, bytes.NewReader(buf[:n]))
-		nwritten += n64
-		if err != nil {
-			return nwritten, err
-		}
+// Open opens a remote DBFS file for reading or writing.
+// The returned object implements relevant [io] interfaces for convenient
+// integration with other code that reads or writes bytes.
+//
+// The [io.WriterTo] interface is provided and maximizes throughput for
+// bulk reads by reading data with the DBFS maximum read chunk size of 1MB.
+// Similarly, the [io.ReaderFrom] interface is provided for bulk writing.
+//
+// A file opened for writing must always be closed.
+func (a DbfsAPI) Open(ctx context.Context, path string, mode FileMode) (Handle, error) {
+	h := &dbfsHandle{
+		ctx:  useragent.InContext(ctx, "sdk-feature", "dbfs-handle"),
+		api:  a,
+		path: path,
+	}
+
+	isRead := (mode & FileModeRead) != 0
+	isWrite := (mode & FileModeWrite) != 0
+	if isRead && isWrite {
+		return nil, fmt.Errorf("dbfs: cannot open file for reading and writing at the same time")
+	}
+
+	var err error
+	switch {
+	case isRead:
+		err = h.openForRead(mode)
+	case isWrite:
+		err = h.openForWrite(mode)
+	default:
+		// No mode specifed. The caller should be explicit so we return an error.
+		return nil, fmt.Errorf("dbfs: must specify dbfs.FileModeRead or dbfs.FileModeWrite")
 	}
+
+	if err != nil {
+		return nil, fmt.Errorf("dbfs: open: %w", err)
+	}
+
+	return h, nil
 }
 
-func (a *DbfsAPI) Open(ctx context.Context, path string) (*FileReader, error) {
-	ctx = useragent.InContext(ctx, "sdk-feature", "dbfs-open")
-	info, err := a.GetStatusByPath(ctx, path)
+// ReadFile is identical to [os.ReadFile] but for DBFS.
+func (a DbfsAPI) ReadFile(ctx context.Context, name string) ([]byte, error) {
+	h, err := a.Open(ctx, name, FileModeRead)
 	if err != nil {
-		return nil, fmt.Errorf("get status: %w", err)
+		return nil, err
 	}
-	return &FileReader{
-		Size: info.FileSize,
-		path: path,
-		ctx:  ctx,
-		api:  a,
-	}, nil
+
+	h_ := h.(*dbfsHandle)
+	buf := make([]byte, h_.reader.size)
+	_, err = h.Read(buf)
+	if err != nil && err != io.EOF {
+		return nil, err
+	}
+	return buf, nil
+}
+
+// WriteFile is identical to [os.WriteFile] but for DBFS.
+func (a DbfsAPI) WriteFile(ctx context.Context, name string, data []byte) error {
+	h, err := a.Open(ctx, name, FileModeWrite|FileModeOverwrite)
+	if err != nil {
+		return err
+	}
+
+	_, err = h.Write(data)
+	cerr := h.Close()
+	if err == nil && cerr != nil {
+		err = cerr
```

Comment:
isn't it ignoring the close error if write error is not nil? :)

Reply:
Yes, intentionally so. It returns the first error that happened.